service/geopmdpy/session.py

```diff
@@ -31,91 +31,563 @@
+                                     input.
+
+            duration (float): The user specified minimum length of time
+                              for the samples to span in units of
+                              seconds.
+
+            period (float): The user specified period between samples
+                            in units of seconds.
+
+            out_stream (file): Object with .write() method where output
+                               will be printed (typically sys.stdout).
+
+        """
+        self._geopm_proxy.PlatformOpenSession()
+        num_period = 0
+        if period != 0:
+            num_period = math.ceil(duration / period)
+        for sample_idx in TimedLoop(num_period, period):
+            signals = self.read_signals(requests)
+            line = self.format_signals(signals, requests.get_formats())
+            out_stream.write(line)
+        self._geopm_proxy.PlatformCloseSession()
+
+    def run_write(self, requests, duration):
+        """Run a write mode session
+
+        Use the GEOPM D-Bus interface to write the requested set of
+        controls.  The control settings will be held for the specified
+        duration of time and the call to this method blocks until the
+        duration of time has elapsed.
+
+        Args:
+            requests (WriteRequests): Request object parsed from user
+                                      input.
+
+            duration (float): Length of time to hold the requested
+                              control settings in units of seconds.
+
+        """
+        self._geopm_proxy.PlatformOpenSession()
+        for rr in requests:
+            self._geopm_proxy.PlatformWriteControl(*rr)
+        time.sleep(duration)
+        self._geopm_proxy.PlatformCloseSession()
+
+    def check_read_args(self, run_time, period):
+        """Check that the run time and period are valid for a read session
+
+        Args:
+            run_time (float): Time duration of the session in seconds.
+
+            period (float): The user specified period between samples
+                            in units of seconds.
+
+        Raises:
+            RuntimeError: The period is greater than the total time or
+                          either is negative.
+
+        """
+        if period > run_time:
+            raise RuntimeError('Specified a period that is greater than the total run time')
+        if period < 0.0 or run_time < 0.0:
+            raise RuntimeError('Specified a negative run time or period')
+
+    def check_write_args(self, run_time, period):
+        """Check that the run time and period are valid for a write session
+
+        Args:
+            run_time (float): Time duration of the session in seconds.
+
+            period (float): The user specified period between samples
+                            in units of seconds.
+
+        Raises:
+            RuntimeError: The period is non-zero or total time is negative.
+
+        """
+        if run_time <= 0.0:
```

Comment:
I'm wondering if we ought to have some sort of private member, call it self._min_write_time or similar, that's the realistic minimum time for which a control can be applied.

This is going to be IO group control latency dependent, DBus latency dependent, etc.  I know it's greater than 0, but beyond that I'm not sure how to put a bound on it.

There should be docu that describes all these latencies and how they relate to this minimum run_time.

Reply:
I'm not sure what someone would be trying to achieve with a small, non-zero value, but it seems like user error if they expect that using a command line tool written in python is going to respond in microseconds.  

If the run_time is very small, we will set the controls, sleep for that period of time and then revert the values.  The latencies you are talking about are captured in the calls to geopm, not in the call to sleep.  We can add some documentation about how the run_time is the duration of time after all controls have been set and just before the process of reverting the values will begin.

Reply:
> We can add some documentation about how the run_time is the duration of time after all controls have been set until just before the process of reverting the values will begin.

This would clear it all up.  I'm just trying to avoid a situation where the time spent to apply the control is significantly greater than the sleep time, and the user being surprised they're not getting a period they ask for.